fms/utils/generation.py

```diff
@@ -1,9 +1,104 @@
-from typing import Any, Callable, List, MutableMapping, Optional, Union
+import inspect
+from typing import Any, Callable, Dict, List, MutableMapping, Optional, Tuple, Union
 
 import torch
+import torch.nn as nn
 import torch.nn.functional as F
 
 
+def __prepare_list_input(
+    input_ids_list: List[torch.Tensor], model: Union[Callable, torch.nn.Module]
+) -> Tuple[torch.Tensor, MutableMapping[str, Any]]:
+    """
+    Convert the list of Tensors to a rectangular tensor. Return extra kwargs for the position_ids and mask, since this
+    will be required to properly handle the rectangular tensor for certain models.
+    """
+    min_len = min([seq.size(0) for seq in input_ids_list])
+    max_len = max([seq.size(0) for seq in input_ids_list])
+
+    params = inspect.signature(
+        model.forward if isinstance(model, nn.Module) else model
+    ).parameters.keys()
+    extra_kwargs = {}
+    needs_mask = "mask" in params and min_len != max_len
+    needs_position_ids = "position_ids" in params and min_len != max_len
+
+    padded_input_ids_list = []
+    mask_list = []
+    position_ids_list = []
+    for input_ids_i in input_ids_list:
+        seq_len = input_ids_i.size(0)
+        pads = torch.zeros(
+            max_len - seq_len, dtype=torch.long, device=input_ids_i.device
+        )
+        non_pads = torch.ones(seq_len, dtype=torch.bool, device=input_ids_i.device)
+
+        # Setting this to 0, however if 0 is the eos, we will end up truncating the output if using truncate_after_eos
+        # once this workflow works for nested tensor, this can probably be removed
+        padded_input_ids_list.append(torch.cat((pads, input_ids_i)))
+
+        # computing this as it's lightweight but could potentially be skipped
+        mask_list.append(torch.cat((pads.bool(), non_pads)))
+        position_ids_list.append(
+            torch.cat(
+                (
+                    pads,
+                    torch.arange(
+                        0, seq_len, dtype=torch.long, device=input_ids_i.device
+                    ),
+                )
+            )
+        )
+
+    input_ids = torch.stack(padded_input_ids_list)
+    if needs_mask:
+        mask = torch.stack(mask_list)
+        mask = (mask.unsqueeze(-1) == mask.unsqueeze(-2)).tril()
+        extra_kwargs["mask"] = mask
+
+    if needs_position_ids:
+        position_ids = torch.stack(position_ids_list)
+        extra_kwargs["position_ids"] = position_ids
+
+    return input_ids, extra_kwargs
+
+
+def __prepare_model_specific_kwargs(
+    iteration: int, use_cache: bool, model_specific_kwargs: MutableMapping[str, Any]
```

Comment:
I think we can do this functional-style and just return the updated kwargs. something like `update_padding_kwargs` might be clearer than `__prepare_model_specific_kwargs`